src/Orleans/Core/ClusterClient.cs

```diff
@@ -0,0 +1,324 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using Orleans.CodeGeneration;
+using Orleans.Runtime;
+using Orleans.Runtime.Configuration;
+using Orleans.Streams;
+
+namespace Orleans
+{
+    /// <summary>
+    /// Client for communicating with clusters of Orleans silos.
+    /// </summary>
+    public class ClusterClient : IInternalClusterClient
+    {
+        private readonly OutsideRuntimeClient runtimeClient;
+        private readonly AsyncLock initLock = new AsyncLock();
+        private LifecycleState state = LifecycleState.Created;
+
+        private enum LifecycleState
+        {
+            Created,
+            Started,
+            Disposed
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ClusterClient"/> class.
+        /// </summary>
+        /// <param name="configuration">The client configuration.</param>
+        public ClusterClient(ClientConfiguration configuration)
+        {
+            if (configuration == null)
+            {
+                throw new ArgumentException(nameof(configuration));
+            }
+
+            this.Configuration = configuration;
+            this.runtimeClient = new OutsideRuntimeClient(configuration);
+        }
+
+        /// <inheritdoc />
+        public bool IsInitialized => this.state == LifecycleState.Started;
+
+        /// <inheritdoc />
+        public IGrainFactory GrainFactory => this.InternalGrainFactory;
+
+        /// <inheritdoc />
+        internal IInternalGrainFactory InternalGrainFactory
+        {
+            get
+            {
+                this.ThrowIfDisposedOrNotInitialized();
+                return this.runtimeClient.InternalGrainFactory;
+            }
+        }
+
+        /// <inheritdoc />
+        public Logger Logger
+        {
+            get
+            {
+                this.ThrowIfDisposedOrNotInitialized();
+                return this.runtimeClient.AppLogger;
+            }
+        }
+
+        /// <inheritdoc />
+        public IServiceProvider ServiceProvider => this.runtimeClient.ServiceProvider;
+
+        /// <inheritdoc />
+        public TimeSpan ResponseTimeout
+        {
+            get { return this.runtimeClient.GetResponseTimeout(); }
+
+            set { this.runtimeClient.SetResponseTimeout(value); }
+        }
+
+        /// <inheritdoc />
+        public Action<InvokeMethodRequest, IGrain> ClientInvokeCallback
+        {
+            get { return this.runtimeClient.ClientInvokeCallback; }
+            set { this.runtimeClient.ClientInvokeCallback = value; }
+        }
```

Comment:
These seem like initialization concerns, so if moving to a separate builder, they should go there